// flatMap-browserify/index.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.flatMap = f()}})(function(){var define,module,exports;return (function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.default = arraySpeciesCreate;
/**
 * @template T
 * @param {T[]} originalArray
 * @param {number} length
 * @returns {T[]}
 */
function arraySpeciesCreate(originalArray, length) {
  var isArray = Array.isArray(originalArray);
  if (!isArray) {
    return Array(length);
  }
  /**
   * @type {ArrayConstructor|undefined|null} C
   */
  var C = Object.getPrototypeOf(originalArray).constructor;
  if (C) {
    // If IsConstructor(C) is true... not sure how this can be reliably checked without invoking it. Likely not insignificant.
    if ((typeof C === 'undefined' ? 'undefined' : _typeof(C)) === 'object' || typeof C === 'function') {
      C = C[Symbol.species.toString()];
      C = C !== null ? C : undefined;
    }
    if (C === undefined) {
      return Array(length);
    }
    if (typeof C !== 'function') {
      throw TypeError('invalid constructor');
    }
    /** @type {Array} */
    var result = new C(length);
    return result;
  }
}

},{}],2:[function(require,module,exports){
'use strict';

var _flattenIntoArray = require('./flatten-into-array');

var _flattenIntoArray2 = _interopRequireDefault(_flattenIntoArray);

var _arraySpeciesCreate = require('./array-species-create');

var _arraySpeciesCreate2 = _interopRequireDefault(_arraySpeciesCreate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

if (!Object.prototype.hasOwnProperty.call(Array.prototype, 'flatMap')) {

  Array.prototype.flatMap = function flatMap(callbackFn, thisArg) {
    var o = Object(this);
    if (!callbackFn || typeof callbackFn.call !== 'function') {
      throw TypeError('callbackFn must be callable.');
    }
    var t = thisArg !== undefined ? thisArg : undefined;

    var a = (0, _arraySpeciesCreate2.default)(o, o.length);
    (0, _flattenIntoArray2.default)(a, o,
    /*start*/0,
    /*depth*/1, callbackFn, t);
    return a.filter(function (x) {
      return x !== undefined;
    }, a);
  };
}

},{"./array-species-create":1,"./flatten-into-array":3}],3:[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = flattenIntoArray;
/**
 * @template T, U
 * @param {Array<U>} target
 * @param {Array<T>} source
 * @param {number} start
 * @param {number} depth
 * @param {function(T): U[]} mapperFunction
 * @param {Array<T>} thisArg
 * @returns {number}
 */
function flattenIntoArray(target, source, start, depth, mapperFunction, thisArg) {

  var mapperFunctionProvied = mapperFunction !== undefined;
  var targetIndex = start;
  var sourceIndex = 0;
  var sourceLen = source.length;
  while (sourceIndex < sourceLen) {
    var p = sourceIndex;
    var exists = !!source[p];
    if (exists === true) {
      var element = source[p];
      if (element) {
        if (mapperFunctionProvied) {
          element = mapperFunction.call(thisArg, element, sourceIndex, target);
        }
        var spreadable = Object.getOwnPropertySymbols(element).includes(Symbol.isConcatSpreadable) || Array.isArray(element);
        if (spreadable === true && depth > 0) {
          var nextIndex = flattenIntoArray(target, element, targetIndex, depth - 1);
          targetIndex = nextIndex;
        } else {
          if (!Number.isSafeInteger(targetIndex)) {
            throw TypeError();
          }
          target[targetIndex] = element;
        }
      }
    }
    targetIndex += 1;
    sourceIndex += 1;
  }
  return targetIndex;
}

},{}],4:[function(require,module,exports){
'use strict';

var _arraySpeciesCreate = require('./array-species-create');

var _arraySpeciesCreate2 = _interopRequireDefault(_arraySpeciesCreate);

var _flattenIntoArray = require('./flatten-into-array');

var _flattenIntoArray2 = _interopRequireDefault(_flattenIntoArray);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

if (!Object.prototype.hasOwnProperty.call(Array.prototype, 'flatten')) {

  /**
   * @param {number=Infinity} depth
   */
  Array.prototype.flatten = function flatten(depth) {
    var o = Object(this);
    var a = (0, _arraySpeciesCreate2.default)(o, this.length);
    var depthNum = depth !== undefined ? Number(depth) : Infinity;
    (0, _flattenIntoArray2.default)(a, o, 0, depthNum);
    return a.filter(function (e) {
      return e !== undefined;
    });
  };
}

},{"./array-species-create":1,"./flatten-into-array":3}],5:[function(require,module,exports){
'use strict';

require('./flatten');

require('./flat-map');

},{"./flat-map":2,"./flatten":4}]},{},[5])(5)
});
